### PR TITLE
Add modules for CP analysis, processing features, and data analysis

### PR DESCRIPTION
**Pull request by mkaykeating**
Modeled everything off of Jenna's cellpainting repo. Module 4 will include different quantification metrics. I started with IoU, although I need to build that out more.

### CODE REVIEW

File: 1.run_multi_IC_pipelines/scripts/multi_pipelines.py [hunk 1]
```diff
@@ -52,16 +44,16 @@
...
+        for name in pipeline_names
 }
 # iterate over the dictionary and add the path_to_pipeline 
 for name, info in plate_info_dictionary.items():
-        for i in range(len(plate_run_names)):
-            if name == "Plate_1_" + pipeline_names[i]:
-                info["path_to_pipeline"] = pathlib.Path(f"../pipelines/" + pipeline_names[i]+ ".cppipe")
+        for i in range(len(pipeline_names)):
+                if name == pipeline_names[i]:
+                    info["path_to_pipeline"] = pathlib.Path(f"../pipelines/" + pipeline_names[i]+ ".cppipe")
```
Comment (MikeLippincott) [inline 1.run_multi_IC_pipelines/scripts/multi_pipelines.py]: consider cleaning up the tabs and formatting

Comment (MikeLippincott) [inline 1.run_multi_IC_pipelines/scripts/multi_pipelines.py]:
```suggestion
        for i in range(len(pipeline_names)):
                if name == pipeline_names[i]:
                    info["path_to_pipeline"] = pathlib.Path(f"../pipelines/{pipeline_names[i]}.cppipe")
```
[f-strings](https://realpython.com/python-f-strings/)  in python are really cool - consider using them more the future for this concept called string interpolation!

File: 2.cellprofiler_analysis/IC_cellprofiler_analysis.sh [hunk 2]
```diff
@@ -0,0 +1,16 @@
+# Modified Jenna's nf1_ic.ipynb file from the Cellpainting repo
+https://github.com/WayScience/nf1_cellpainting_data/blob/main/2.cellprofiler_analysis/nf1_analysis.sh
```
Comment (MikeLippincott) [inline 2.cellprofiler_analysis/IC_cellprofiler_analysis.sh]:
```suggestion
# https://github.com/WayScience/nf1_cellpainting_data/blob/main/2.cellprofiler_analysis/nf1_analysis.sh
```
consider commenting this out

File: 2.cellprofiler_analysis/IC_cellprofiler_analysis.sh [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+
+#!/bin/bash
```
Comment (MikeLippincott) [inline 2.cellprofiler_analysis/IC_cellprofiler_analysis.sh:1]: Also this line will need to moved to the first line - best practice. This is the line that tells the machine to read this as a bash script

File: 3.processing_features/notebooks/runinfo/000/submit_scripts/parsl.htex_default_for_cytotable.block-0.1705531433.3331797.sh
Comment (MikeLippincott) [inline 3.processing_features/notebooks/runinfo/000/submit_scripts/parsl.htex_default_for_cytotable.block-0.1705531433.3331797.sh:1]: There is a lot going on in the script. Please write comments for each line to help yourself and the reader/user

File: 3.processing_features/notebooks/runinfo/000/submit_scripts/parsl.htex_default_for_cytotable.block-0.1705531433.3331797.sh.ec
Comment (MikeLippincott) [inline 3.processing_features/notebooks/runinfo/000/submit_scripts/parsl.htex_default_for_cytotable.block-0.1705531433.3331797.sh.ec:1]: consider removing or `.gitignoring this file`

File: 3.processing_features/notebooks/runinfo/000/submit_scripts/parsl.htex_default_for_cytotable.block-0.1705531433.3331797.sh.err
Comment (MikeLippincott) [inline 3.processing_features/notebooks/runinfo/000/submit_scripts/parsl.htex_default_for_cytotable.block-0.1705531433.3331797.sh.err:1]: consider a `.gitignore` here too

File: 3.processing_features/notebooks/runinfo/000/submit_scripts/parsl.htex_default_for_cytotable.block-0.1705531433.3331797.sh.out
Comment (MikeLippincott) [inline 3.processing_features/notebooks/runinfo/000/submit_scripts/parsl.htex_default_for_cytotable.block-0.1705531433.3331797.sh.out:1]: here too

File: 3.processing_features/scripts/0.merge_sc_cytotable.py [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+# https://github.com/WayScience/nf1_cellpainting_data/blob/main/3.processing_features/0.merge_sc_cytotable.ipynb
+
+import sys
+import pathlib
+import yaml
+import pprint
+import pandas as pd
+
+# cytotable will merge objects from SQLite file into single cells and save as parquet file
+from cytotable import convert, presets
```
Comment (MikeLippincott) [inline 3.processing_features/scripts/0.merge_sc_cytotable.py]: consider changing the import order: [here](https://peps.python.org/pep-0008/#imports)

File: 3.processing_features/scripts/0.merge_sc_cytotable.py [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+for file_path in pathlib.Path("../../2.cellprofiler_analysis/pipelines/IC_function_apply_pipelines").iterdir():
+    if str(file_path.stem).startswith("IC_function"):
+        pipeline_names.append(str(file_path.stem))
+
+print(pipeline_names)
+
+
+# ## Create dictionary with info for each plate
+# 
+# **Note:** All paths must be string to use with CytoTable.
```
Comment (MikeLippincott) [inline 3.processing_features/scripts/0.merge_sc_cytotable.py:65]: nice documentation

File: 3.processing_features/scripts/0.merge_sc_cytotable.py [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+        "dest_path": str(pathlib.Path(f"{output_dir}/{name}.parquet")),
+    }
+    for name in pipeline_names 
+}
+
+# iterate over the dictionary and add the source_path
+for name, info in plate_info_dictionary.items():
+        for i in range(len(pipeline_names)):
+            if name == pipeline_names[i]:
+                  info["source_path"] = str(pathlib.Path(f"../../2.cellprofiler_analysis/outputs/SQLites/" + pipeline_names[i]+ ".sqlite"))
```
Comment (MikeLippincott) [inline 3.processing_features/scripts/0.merge_sc_cytotable.py]: please see above comment about fstrings :)

File: 4.analyze_data/README.md
Comment (MikeLippincott) [inline 4.analyze_data/README.md:1]: consider removing if empty

File: environments/IC_bench_4.2.6.yml [hunk 7]
```diff
@@ -14,9 +14,12 @@ dependencies:
   - conda-forge::pathlib
   - conda-forge::nbconvert
   - conda-forge::Cython==0.29.36
+  - conda-forge/label/broken::pytorch
```
Comment (MikeLippincott) [inline environments/IC_bench_4.2.6.yml]: label/broken???

File: 4.analyze_data/IoU.ipynb
Comment (MikeLippincott) [inline 4.analyze_data/IoU.ipynb:1]: I am not seeing a `.py` file here. Please consider adding one.

File: 4.analyze_data/IoU.ipynb [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "tensor(1.)\n"
+     ]
+    }
+   ],
+   "source": [
+    "# Intersection over Union function (IoU) utilizing ground truth (gt) and test binary masks \n",
+    "def iou(gt, test):\n",
```
Comment (MikeLippincott) [inline 4.analyze_data/IoU.ipynb]: consider adding a docstring here and type hints

File: 4.analyze_data/IoU.ipynb [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+    "\n",
+    "# Read binary masks \n",
+    "mask1 = cv2.imread('/home/maggiekeating/M1_B1_01_1_2_DAPI_001_nuclei_mask.tiff', 0)\n",
+    "mask2 = cv2.imread('/home/maggiekeating/M2_B1_01_1_2_DAPI_001_nuclei_mask.tiff', 0)\n",
+    "\n",
+    "# Convert binary masks from numpy ndarray to pytorch tensor \n",
+    "mask1_conv = torch.from_numpy(mask1)\n",
+    "mask2_conv = torch.from_numpy(mask2)\n",
+    "\n",
+    "print(iou(mask1_conv, mask2_conv))"
```
Comment (MikeLippincott) [inline 4.analyze_data/IoU.ipynb]: might be on your to-do list but consider outputting these iou stats for downstream analysis

**Review by MikeLippincott — CHANGES_REQUESTED** [review]
Looks really good and has some great stuff going on here. Please read and address comments. Needs a bit more cleaning up and then ready to merge soon!

File: 1.run_multi_IC_pipelines/scripts/multi_pipelines.py (hunk 1 above)
Comment (mkaykeating) [reply 1.run_multi_IC_pipelines/scripts/multi_pipelines.py]: Got it, thanks!

File: 2.cellprofiler_analysis/IC_cellprofiler_analysis.sh (hunk 3 above)
Comment (mkaykeating) [reply 2.cellprofiler_analysis/IC_cellprofiler_analysis.sh:1]: Thanks for letting me know! I updated all of my sh files

File: 3.processing_features/notebooks/runinfo/000/submit_scripts/parsl.htex_default_for_cytotable.block-0.1705531433.3331797.sh
Comment (mkaykeating) [reply 3.processing_features/notebooks/runinfo/000/submit_scripts/parsl.htex_default_for_cytotable.block-0.1705531433.3331797.sh:1]: This sh file as well as the others in this folder are generated when running the scripts in this notebook. Should I just use gitignore for all 4 of these files in the submit_scripts folder?

File: 3.processing_features/scripts/0.merge_sc_cytotable.py (hunk 4 above)
Comment (mkaykeating) [reply 3.processing_features/scripts/0.merge_sc_cytotable.py]: I believe I changed it to the correct order now. If you agree, I will modify all of my other scripts as well

File: 3.processing_features/scripts/0.merge_sc_cytotable.py (hunk 5 above)
Comment (mkaykeating) [reply 3.processing_features/scripts/0.merge_sc_cytotable.py:65]: I have Jenna to thank for that :)

File: 3.processing_features/scripts/0.merge_sc_cytotable.py (hunk 6 above)
Comment (mkaykeating) [reply 3.processing_features/scripts/0.merge_sc_cytotable.py]: Got it, thank you for showing me this!

File: 4.analyze_data/README.md
Comment (mkaykeating) [reply 4.analyze_data/README.md:1]: Done! This module is super incomplete atm

File: environments/IC_bench_4.2.6.yml (hunk 7 above)
Comment (mkaykeating) [reply environments/IC_bench_4.2.6.yml]: Lol anaconda.org said use one of the following 1. conda install conda-forge::pytorch
2. conda install conda-forge/label/broken::pytorch and I impulsively picked the weird one. I switched it to conda install conda-forge::pytorch. Thanks for pointing  this out.

File: 4.analyze_data/IoU.ipynb (hunk 8 above)
Comment (mkaykeating) [reply 4.analyze_data/IoU.ipynb]: Got it :) Thanks

File: environments/IC_bench_4.2.6.yml (hunk 7 above)
Comment (mkaykeating) [reply environments/IC_bench_4.2.6.yml:17]: Whoops fixed it

File: 4.analyze_data/IoU.ipynb (hunk 9 above)
Comment (mkaykeating) [reply 4.analyze_data/IoU.ipynb]: I was thinking about outputting them to a CSV. Is that a good way to go about it?

File: 4.analyze_data/IoU.ipynb
Comment (mkaykeating) [reply 4.analyze_data/IoU.ipynb:1]: Missed that, thanks!!

File: 1.run_multi_IC_pipelines/data/Metadata/platemap_NF1_plate3_sub.csv
Comment (jenna-tomkinson) [inline 1.run_multi_IC_pipelines/data/Metadata/platemap_NF1_plate3_sub.csv:1]: Do you need this extra platemap file or can this be removed? Just curious!

File: 1.run_multi_IC_pipelines/scripts/multi_pipelines.py
Comment (jenna-tomkinson) [inline 1.run_multi_IC_pipelines/scripts/multi_pipelines.py:1]: Love that you made the paths less strict and relative to the repo! Great job!

File: 3.processing_features/processing_features.sh [hunk 10]
```diff
@@ -0,0 +1,19 @@
...
+conda activate IC_bench_4.2.6
+
+# convert all notebooks to python files into the scripts folder
+jupyter nbconvert --to python --output-dir=scripts/ *.ipynb
+
+# run each python script
+python scripts/0.merge_sc_cytotable.py
+python scripts/1.pycytominer_bulk_pipelines.py
+python scripts/2.pycytominer_singlecell_pipelines.py
+
```
Comment (jenna-tomkinson) [inline 3.processing_features/processing_features.sh]: Recommend adding new endline here to remove error (add extra empty line to bottom of files)

**Review by jenna-tomkinson — APPROVED** [review]
Amazing job implementing everything! My only small comment is you might what to update the dictionary name in notebooks from `plate_info_dictionary` to maybe `pipeline_info` or something like that since you are only using one plate.

Great job and looks ready to merge after addressing one comment I do believe I left!

File: 1.run_multi_IC_pipelines/data/Metadata/platemap_NF1_plate3_sub.csv
Comment (mkaykeating) [reply 1.run_multi_IC_pipelines/data/Metadata/platemap_NF1_plate3_sub.csv:1]: Good catch! I don't need that, I will remove.

File: 1.run_multi_IC_pipelines/scripts/multi_pipelines.py
Comment (mkaykeating) [reply 1.run_multi_IC_pipelines/scripts/multi_pipelines.py:1]: Thanks!!!

File: 3.processing_features/processing_features.sh (hunk 10 above)
Comment (mkaykeating) [reply 3.processing_features/processing_features.sh]: Got it, thanks!

**Comment by mkaykeating** [comment]
> Amazing job implementing everything! My only small comment is you might what to update the dictionary name in notebooks from `plate_info_dictionary` to maybe `pipeline_info` or something like that since you are only using one plate.
> 
> Great job and looks ready to merge after addressing one comment I do believe I left!

F

> Amazing job implementing everything! My only small comment is you might what to update the dictionary name in notebooks from `plate_info_dictionary` to maybe `pipeline_info` or something like that since you are only using one plate.
> 
> Great job and looks ready to merge after addressing one comment I do believe I left!

Thanks! And good idea! I swapped 'plate_info_dictionary' with pipeline_info_dictionary'

**Pull request merged**